src/libsyntax/ext/expand.rs

```diff
@@ -640,6 +640,9 @@ pub fn expand_item_mac(it: P<ast::Item>, fld: &mut MacroExpander)
             // need to be marked. Not that it could be marked anyway.
             // create issue to recommend refactoring here?
             fld.cx.syntax_env.insert(intern(name[]), ext);
+            // Keep track of this macro definition both within the crate
+            // context and (if applicable) in its exports.
+            fld.cx.macros.push(it.clone());
```

Comment:
I don't think that this quite catches shadowing per-se because you could have two disjoint modules with the same macro with the same name, but they're not shadowing one another.

Where you had the logic before I believe was the correct place to put the logic. Did the `add_lint` method not work out?